File: test/65_settings.js
var assert = require('assert');

var helper = require('./helper'),
    protocol = helper.protocol;

describe('6.5. SETTINGS', function () {
  it('Sends a SETTINGS frame', function (done) {
    var message = 'the endpoint MUST sends a SETTINGS frame with ACK.';

    var conn = helper.createConnection(function () {
      conn.once('frame', function (frame) {
        if (frame.type === protocol.FRAME_TYPE_SETTINGS) {
          assert.ok(frame.ack, message);
          done();
        }
      });

      var frame = helper.createSettingsFrame();
      frame.setMaxConcurrentStreams(100);
      conn.socket.write(frame.encode());
    });
  });

  it('Sends a SETTINGS frame that is not a zero-length with ACK flag', function (done) {
    var message = 'the endpoint MUST respond with a connection error of type FRAME_SIZE_ERROR.';

    var conn = helper.createConnection(function () {
      conn.once('error', function (err) {
        assert.equal(err.code, protocol.CODE_FRAME_SIZE_ERROR, message);
        done();
      });

      var frame = helper.createSettingsFrame();
      frame.length = 1;
      frame.ack = true;
      conn.socket.write(frame.encode());
    });
  });

  it('Sends a SETTINGS frame with the stream identifier that is not 0x0', function (done) {
    var message = 'the endpoint MUST respond with a connection error of type PROTOCOL_ERROR.';

    var conn = helper.createConnection(function () {
      conn.once('error', function (err) {
        assert.equal(err.code, protocol.CODE_PROTOCOL_ERROR, message);
        done();
      });

      var frame = helper.createSettingsFrame();
      frame.streamId = 0x3;
      frame.setMaxConcurrentStreams(100);
      conn.socket.write(frame.encode());
    });
  });

  it('Sends a incomplete SETTINGS frame', function (done) {
    var message = 'the endpoint MUST respond with a connection error of type PROTOCOL_ERROR.';

    var conn = helper.createConnection(function () {
      conn.once('error', function (err) {
        assert.equal(err.code, protocol.CODE_PROTOCOL_ERROR, message);
        done();
      });

      var frame = new Buffer([
        0x00, 0x00, 0x02, 0x04, 0x04, 0x00, 0x00, 0x00,
        0x00, 0x00, 0x01, 0x00, 0x00, 0x00, 0x64
      ]);
      conn.socket.write(frame);
    });
  });

  describe('6.5.2.  Defined SETTINGS Parameters', function () {
    describe('SETTINGS_ENABLE_PUSH (0x2)', function () {
      it('Sends the value other than 0 or 1', function (done) {
        var message = 'the endpoint MUST respond with a connection error of type PROTOCOL_ERROR.';

        var conn = helper.createConnection(function () {
          conn.once('error', function (err) {
            assert.equal(err.code, protocol.CODE_PROTOCOL_ERROR, message);
            done();
          });

          var frame = new Buffer([
            0x00, 0x00, 0x06, 0x04, 0x00, 0x00, 0x00, 0x00,
            0x00, 0x00, 0x02, 0x00, 0x00, 0x00, 0x02
          ]);
          conn.socket.write(frame);
        });
      });
    });

    describe('SETTINGS_INITIAL_WINDOW_SIZE (0x4)', function () {
      it('Sends the value above the maximum flow control window size', function (done) {
        var message = 'the endpoint MUST respond with a connection error of type FLOW_CONTROL_ERROR.';

        var conn = helper.createConnection(function () {
          conn.once('error', function (err) {
            assert.equal(err.code, protocol.CODE_FLOW_CONTROL_ERROR, message);
            done();
          });

          var frame = new Buffer([
            0x00, 0x00, 0x06, 0x04, 0x00, 0x00, 0x00, 0x00,
            0x00, 0x00, 0x04, 0x80, 0x00, 0x00, 0x00
          ]);
          conn.socket.write(frame);
        });
      });
    });

    describe('SETTINGS_MAX_FRAME_SIZE (0x5)', function () {
      it('Sends the value below the initial value', function (done) {
        var message = 'the endpoint MUST respond with a connection error of type PROTOCOL_ERROR.';

        var conn = helper.createConnection(function () {
          conn.once('error', function (err) {
            assert.equal(err.code, protocol.CODE_PROTOCOL_ERROR, message);
            done();
          });

          var frame = new Buffer([
            0x00, 0x00, 0x06, 0x04, 0x00, 0x00, 0x00, 0x00,
            0x00, 0x00, 0x05, 0x00, 0x00, 0x3f, 0xff
          ]);
          conn.socket.write(frame);
        });
      });

      it('Sends the value above the maximum allowed frame size', function (done) {
        var message = 'the endpoint MUST respond with a connection error of type PROTOCOL_ERROR.';

        var conn = helper.createConnection(function () {
          conn.once('error', function (err) {
            assert.equal(err.code, protocol.CODE_PROTOCOL_ERROR, message);
            done();
          });

          var frame = new Buffer([
            0x00, 0x00, 0x06, 0x04, 0x00, 0x00, 0x00, 0x00,
            0x00, 0x00, 0x05, 0x01, 0x00, 0x00, 0x00
          ]);
          conn.socket.write(frame);
        });
      });
    });
  });
});

File: client/6_5_settings.go
package client

import (
	"errors"
	"golang.org/x/net/http2"

	"github.com/summerwind/h2spec/config"
	"github.com/summerwind/h2spec/spec"
)

func Settings() *spec.ClientTestGroup {
	tg := NewTestGroup("6.5", "SETTINGS")

	// ACK (0x1):
	// When set, bit 0 indicates that this frame acknowledges receipt
	// and application of the peer's SETTINGS frame. When this bit is
	// set, the payload of the SETTINGS frame MUST be empty. Receipt of
	// a SETTINGS frame with the ACK flag set and a length field value
	// other than 0 MUST be treated as a connection error (Section 5.4.1)
	// of type FRAME_SIZE_ERROR.
	tg.AddTestCase(&spec.ClientTestCase{
		Desc:        "Sends a SETTINGS frame with ACK flag and payload",
		Requirement: "The endpoint MUST respond with a connection error of type FRAME_SIZE_ERROR.",
		Run: func(c *config.Config, conn *spec.Conn) error {
			_, err := conn.ReadClientPreface()
			if err != nil {
				return err
			}

			_, ok := conn.WaitEventByType(spec.EventSettingsFrame)
			if !ok {
				return errors.New("First frame from client must be SETTINGS")
			}

			// SETTINGS frame:
			// length: 0, flags: 0x1, stream_id: 0x0
			conn.Send([]byte("\x00\x00\x01\x04\x01\x00\x00\x00\x00"))
			conn.Send([]byte("\x00"))

			return spec.VerifyConnectionError(conn, http2.ErrCodeFrameSize)
		},
	})

	// SETTINGS frames always apply to a connection, never a single
	// stream. The stream identifier for a SETTINGS frame MUST be
	// zero (0x0). If an endpoint receives a SETTINGS frame whose
	// stream identifier field is anything other than 0x0, the
	// endpoint MUST respond with a connection error (Section 5.4.1)
	// of type PROTOCOL_ERROR.
	tg.AddTestCase(&spec.ClientTestCase{
		Desc:        "Sends a SETTINGS frame with a stream identifier other than 0x0",
		Requirement: "The endpoint MUST respond with a connection error of type PROTOCOL_ERROR.",
		Run: func(c *config.Config, conn *spec.Conn) error {
			_, err := conn.ReadClientPreface()
			if err != nil {
				return err
			}

			_, ok := conn.WaitEventByType(spec.EventSettingsFrame)
			if !ok {
				return errors.New("First frame from client must be SETTINGS")
			}

			// SETTINGS frame:
			// length: 6, flags: 0x0, stream_id: 0x1
			conn.Send([]byte("\x00\x00\x06\x04\x00\x00\x00\x00\x01"))
			conn.Send([]byte("\x00\x03\x00\x00\x00\x64"))

			return spec.VerifyConnectionError(conn, http2.ErrCodeProtocol)
		},
	})

	// The SETTINGS frame affects connection state. A badly formed or
	// incomplete SETTINGS frame MUST be treated as a connection error
	// (Section 5.4.1) of type PROTOCOL_ERROR.
	//
	// A SETTINGS frame with a length other than a multiple of 6 octets
	// MUST be treated as a connection error (Section 5.4.1) of type
	// FRAME_SIZE_ERROR.
	tg.AddTestCase(&spec.ClientTestCase{
		Desc:        "Sends a SETTINGS frame with a length other than a multiple of 6 octets",
		Requirement: "The endpoint MUST respond with a connection error of type FRAME_SIZE_ERROR.",
		Run: func(c *config.Config, conn *spec.Conn) error {
			_, err := conn.ReadClientPreface()
			if err != nil {
				return err
			}

			_, ok := conn.WaitEventByType(spec.EventSettingsFrame)
			if !ok {
				return errors.New("First frame from client must be SETTINGS")
			}

			// SETTINGS frame:
			// length: 3, flags: 0x0, stream_id: 0x0
			conn.Send([]byte("\x00\x00\x03\x04\x00\x00\x00\x00\x00"))
			conn.Send([]byte("\x00\x03\x00"))

			codes := []http2.ErrCode{
				http2.ErrCodeProtocol,
				http2.ErrCodeFrameSize,
			}
			return spec.VerifyStreamError(conn, codes...)
		},
	})

	tg.AddTestGroup(DefinedSETTINGSParameters())
	tg.AddTestGroup(SettingsSynchronization())

	return tg
}
